File: src/150. Evaluate Reverse Polish Notation/main.ts
// Solved by @kitanoyoru
// https://leetcode.com/problems/evaluate-reverse-polish-notation/

const action = (firstOp: string, temp: string, lastOp: string) => {
  switch (temp) {
    case "+":
      console.log(firstOp, lastOp)
      return String(parseInt(firstOp) + parseInt(lastOp))
    case "-":
      return String(parseInt(firstOp) - parseInt(lastOp))
    case "/":
      return String(parseInt(firstOp) / parseInt(lastOp))
    case "*":
      return String(parseInt(firstOp) * parseInt(lastOp))
  }
}

const evalRPN = (tokens: string[]) => {
  let stack: string[] = []

  for (let i = 0; i < tokens.length; i++) {
    if (Number(tokens[i])) {
      stack.push(tokens[i])
    } else {
      let lastOp = stack.pop()!
      let firstOp = stack.pop()!
      console.log(firstOp, tokens[i], lastOp)
      stack.push(action(firstOp, tokens[i], lastOp)!)
    }
  }

  return stack[0]
}

console.log(evalRPN(["4", "-2", "/", "2", "-3", "-", "-"]))
console.log(evalRPN(["4", "13", "5", "/", "+"]))

File: src/150. Evaluate Reverse Polish Notation/main.go
package main

import "strconv"

type Stack []int

func NewStack() *Stack {
	return &Stack{}
}

func (s *Stack) Push(v int) {
	*s = append(*s, v)
}

func (s *Stack) Pop() int {
	res := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]

	return res
}

func (s *Stack) Peek() int {
	return (*s)[0]
}

func evalRPN(tokens []string) int {
	s := NewStack()
	for _, token := range tokens {
		if token == "+" {
			s.Push(s.Pop() + s.Pop())
		} else if token == "-" {
			temp := s.Pop()
			s.Push(s.Pop() - temp)
		} else if token == "*" {
			s.Push(s.Pop() * s.Pop())
		} else if token == "/" {
			temp := s.Pop()
			s.Push(s.Pop() / temp)
		} else {
			v, _ := strconv.Atoi(token)
			s.Push(v)
		}
	}

	return s.Peek()
}
